### PR TITLE
SONARPY-764 Rule S5914: Assertions should not fail or succeed unconditionally

### PR DESCRIPTION
**Pull request by nils-werner-sonarsource**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/CheckUtils.java [hunk 1]
```diff
@@ -104,4 +119,49 @@ private static boolean calleeHasNameLocals(CallExpression callExpression) {
...
+      return true;
+    }
+    if (symbol.is(Symbol.Kind.FUNCTION)) {
+      // Avoid potential FPs with properties: only report on limited selection of "safe" decorators
+      return ACCEPTED_DECORATORS.containsAll(((FunctionSymbol) symbol).decorators());
+    }
+    return false;
+  }
+
+  private static boolean isAlwaysEmptyOrNonEmptyCollection(List<? extends Tree> elements) {
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/CheckUtils.java:161]: This method name also feels kind of odd - I know you didn't choose it, but it strikes me a bit now that it's in `CheckUtils`.
Probably `doesNotContainUnpackingExpression` is a bit more generic.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
I'm sorry I won't have time to complete the review today, but I already left some comments that I think are worth addressing. We can of course discuss them tomorrow :)

File: its/ruling/src/test/resources/expected/python-S5914.json [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+],
+'project:numpy-1.16.4/tools/npy_tempita/__init__.py':[
+272,
+1029,
+],
+'project:tensorflow/python/client/events_writer_test.py':[
+51,
+],
+'project:tensorflow/python/data/experimental/kernel_tests/group_by_window_test.py':[
+222,
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply its/ruling/src/test/resources/expected/python-S5914.json:13]: It's a generator expression. Not sure how to handle this in Python.

File: its/ruling/src/test/resources/expected/python-S5914.json [hunk 3]
```diff
@@ -0,0 +1,121 @@
+{
+'project:biopython/Bio/PDB/DSSP.py':[
+194,
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply its/ruling/src/test/resources/expected/python-S5914.json]: Fixed by 59ffcc2f9582face247f0d41d0a06ea91bfb7a63

File: its/ruling/src/test/resources/expected/python-S5914.json [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+219,
+],
+'project:tornado-2.3/tornado/test/twisted_test.py':[
+400,
+],
+'project:twisted-12.1.0/twisted/conch/ssh/sexpy.py':[
+31,
+],
+'project:twisted-12.1.0/twisted/internet/test/connectionmixins.py':[
+482,
```
[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply its/ruling/src/test/resources/expected/python-S5914.json]: Fixed by 59ffcc2f9582face247f0d41d0a06ea91bfb7a63

File: its/ruling/src/test/resources/expected/python-S5914.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (guillaume-dequenne-sonarsource) [reply its/ruling/src/test/resources/expected/python-S5914.json:13]: I see, I was surprised at the lack of parentheses (outside of the assert argument, the expression needs to be parenthesized). It's a quite nice TP actually! The expression returns a generator no matter the size or truthiness of the list.

[the remark this answers is not in the document]
Comment (nils-werner-sonarsource) [reply its/ruling/src/test/resources/expected/python-S5914.json:13]: Whoop Whoop

File: python-checks/src/main/java/org/sonar/python/checks/tests/UnconditionalAssertionCheck.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      List<Argument> arguments = call.arguments();
+
+      if (BOOLEAN_ASSERTIONS.contains(name)) {
+        checkBooleanAssertion(ctx, TreeUtils.nthArgumentOrKeyword(0, "testValue", arguments));
+      } else if (NONE_ASSERTIONS.contains(name)) {
+        checkNoneAssertion(ctx, call, TreeUtils.nthArgumentOrKeyword(0, "testValue", arguments));
+      } else if (IS_ASSERTIONS.contains(name)) {
+        String message = "assertIs".equals(name) ? IS_MESSAGE : IS_NOT_MESSAGE;
+        checkIsAssertion(ctx, call, TreeUtils.nthArgumentOrKeyword(0, "firstValue", arguments), message);
+        checkIsAssertion(ctx, call, TreeUtils.nthArgumentOrKeyword(1, "secondValue", arguments), message);
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/tests/UnconditionalAssertionCheck.java:92]: Where did you see the keyword is `firstValue` and `secondValue`? In my `unittest.TestCase`, I have `expr1` and `expr2` as keywords (Python 3.10).

File: python-checks/src/main/java/org/sonar/python/checks/tests/UnconditionalAssertionCheck.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        return;
+      }
+
+      String name = symbol.name();
+      List<Argument> arguments = call.arguments();
+
+      if (BOOLEAN_ASSERTIONS.contains(name)) {
+        checkBooleanAssertion(ctx, TreeUtils.nthArgumentOrKeyword(0, "testValue", arguments));
+      } else if (NONE_ASSERTIONS.contains(name)) {
+        checkNoneAssertion(ctx, call, TreeUtils.nthArgumentOrKeyword(0, "testValue", arguments));
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/tests/UnconditionalAssertionCheck.java:88]: Here I have `obj` as keyword.

File: python-checks/src/main/java/org/sonar/python/checks/tests/UnconditionalAssertionCheck.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      Symbol symbol = call.calleeSymbol();
+      if (symbol == null) {
+        return;
+      }
+
+      String name = symbol.name();
+      List<Argument> arguments = call.arguments();
+
+      if (BOOLEAN_ASSERTIONS.contains(name)) {
+        checkBooleanAssertion(ctx, TreeUtils.nthArgumentOrKeyword(0, "testValue", arguments));
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/tests/UnconditionalAssertionCheck.java:86]: I have `expr` as keyword locally.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Results are looking good and implementation as well. There are still a few comments from my previous review that were not addressed (I don't really like that we introduce logic specific to a couple of rules in `CheckUtils` class).
Also, there is something weird with keyword arguments for the unittest assertions.
The documentation [here](https://docs.python.org/3.10/library/unittest.html#assert-methods) provides the list of methods and their keyword arguments, which seems to be different from what you use. Weirdly enough, in my local test project, the keywords arguments are also different from both your version and the documentation...If we can't figure why that is, I believe we should default to what is in the documentation.

**Pull request merged**